### PR TITLE
Adds DeprecatedHelperVisitor

### PR DESCRIPTION
**Pull request by bbrala**
Adds DeprecatedHelperVisitor which removes PHPStan parsing of everything in the 3rd argument of the DeprecationHelper, making it ignore deprecation messages in there.

### CODE REVIEW

File: src/Parser/DeprecatedHelperVisitor.php [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+class DeprecatedHelperVisitor extends NodeVisitorAbstract implements NodeVisitor
+{
+
+    private ?Node\Arg $deprecatedCall = null;
+
+    public function enterNode(Node $node)
+    {
+
+        if ($node instanceof Node\Expr\StaticCall) {
+            if ($node->class instanceof Node\Name\FullyQualified && $node->class->isFullyQualified() && $node->class->toString() === 'Drupal\Component\Utility\DeprecationHelper') {
```
Comment (mglaman) [inline src/Parser/DeprecatedHelperVisitor.php]: It's always an instance of FullyQualified even if imported, right? Pretty sure but double checking.

File: src/Parser/DeprecatedHelperVisitor.php [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+            }
+        }
+
+        if ($this->deprecatedCall !== null && $node instanceof Node\Arg && $this->isSavedArgument($node)) {
+            $this->deprecatedCall = null;
+            return NodeTraverser::DONT_TRAVERSE_CURRENT_AND_CHILDREN;
+        }
+
+        return null;
+    }
```
Comment (mglaman) [inline src/Parser/DeprecatedHelperVisitor.php:30]: Is there a chance our property won't get unset for some reason?

File: src/Parser/DeprecatedHelperVisitor.php [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    {
+        if ($node instanceof Node\Expr\StaticCall) {
+            if ($node->class instanceof Node\Name\FullyQualified && $node->class->toString() === 'Drupal\Component\Utility\DeprecationHelper') {
+                $this->deprecatedCall = $node->getArgs()[2];
+                return null;
+            }
+        }
+
+        if ($this->deprecatedCall !== null && $node instanceof Node\Arg && $this->isSavedArgument($node)) {
+            $this->deprecatedCall = null;
```
Comment (mglaman) [inline src/Parser/DeprecatedHelperVisitor.php:25]: The null check is technically duplicated in the `isSavedArgument` method

File: src/Parser/DeprecatedHelperVisitor.php [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+            return NodeTraverser::DONT_TRAVERSE_CURRENT_AND_CHILDREN;
+        }
+
+        return null;
+    }
+
+    private function isSavedArgument(Node\Arg $node): bool
+    {
+        if ($this->deprecatedCall !== null && $node->getAttributes() === $this->deprecatedCall->getAttributes()) {
+            return true;
```
Comment (mglaman) [inline src/Parser/DeprecatedHelperVisitor.php:35]: Okay so when we detect a static call on the methods, we fetch and stash the callable for the deprecated case. Then when then node parser reaches the argument we check and unset.

File: tests/src/Parser/DeprecatedHelperVisitor/data/deprecation-helper-test.php [hunk 5]
```diff
@@ -0,0 +1,26 @@
+<?php
+
+namespace GroupLegacy;
```
Comment (bbrala) [inline tests/src/Parser/DeprecatedHelperVisitor/data/deprecation-helper-test.php:3]: This should be a more sane namespace i guess

File: src/Parser/DeprecatedHelperVisitor.php [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+            if ($node->class instanceof Node\Name\FullyQualified && $node->class->toString() === 'Drupal\Component\Utility\DeprecationHelper') {
+                $this->deprecatedCall = $node->getArgs()[2];
+                return null;
+            }
+        }
+
+        if ($this->deprecatedCall !== null && $node instanceof Node\Arg && $this->isSavedArgument($node)) {
+            $this->deprecatedCall = null;
+            return NodeTraverser::DONT_TRAVERSE_CURRENT_AND_CHILDREN;
+        }
```
Comment (mglaman) [inline src/Parser/DeprecatedHelperVisitor.php:27]: This is the magic, yeah? PHPStan doesn't analyze anything in the tree? Seems problematic and instead we should try adding information if possible. Probably not, though.

File: src/Parser/DeprecatedHelperVisitor.php (hunk 1 above)
Comment (bbrala) [reply src/Parser/DeprecatedHelperVisitor.php]: yeah i removed this one after your review

File: src/Parser/DeprecatedHelperVisitor.php (hunk 2 above)
Comment (bbrala) [reply src/Parser/DeprecatedHelperVisitor.php:30]: Calling deprecation helper without arguments? But that is impossible. 

Perhaps if while moving into the function something changes the node, so the argument is no longer on the same line and therefor not equal? Not sure how, and if that is possible. 

But the effect would basically be if would check the array a few times extra, which should be quite cheap.

There might be a way to add some more checks when in the argument. Or do something like keeping a stack of the last static call and check if that call is nothing weird. 

```
        if ($node instanceof Node\Expr\StaticCall) {
            $this->staticCall = $node; // this and check this?
            if ($node->class instanceof Node\Name\FullyQualified && $node->class->toString() === 'Drupal\Component\Utility\DeprecationHelper') {
                $this->deprecatedCall = $node->getArgs()[2];
                return null;
            }
        }
```

But that seems like overkill.

File: src/Parser/DeprecatedHelperVisitor.php (hunk 3 above)
Comment (bbrala) [reply src/Parser/DeprecatedHelperVisitor.php:25]: I was making phpstan happy, the property can be null.

File: src/Parser/DeprecatedHelperVisitor.php (hunk 4 above)
Comment (bbrala) [reply src/Parser/DeprecatedHelperVisitor.php:35]: yes!

File: src/Parser/DeprecatedHelperVisitor.php (hunk 6 above)
Comment (bbrala) [reply src/Parser/DeprecatedHelperVisitor.php:27]: Yeah, it stops parsing as discussed on slack.

File: tests/src/Parser/DeprecatedHelperVisitor/data/deprecation-helper-test.php (hunk 5 above)
Comment (mglaman) [reply tests/src/Parser/DeprecatedHelperVisitor/data/deprecation-helper-test.php:3]: It's good enough for the test data :)

**Comment by bbrala** [comment]
Hi @mglaman , any chance you can revisit this? I understand the drawbacks, but we need some solution to skip code inside the deprecations helper before we start posting patches.

**Comment by mglaman** [comment]
:/ sorry it's been in my TODOs



I'll get to this shortly. I'm thinking we can enable it behind a feature flag. The project update bot can opt in. I'm still concerned that it just halts analysis.

**Comment by bbrala** [comment]
This has been superceeded by #714, yay for phpstan developer @ondrejmirtes (and you @mglaman ;))

**Pull request closed without merging**